torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,734 @@
+
+    NCCL_CHECK(ncclReduce(data[i].data_ptr(),
+                          data[i].data_ptr(),
+                          data[i].numel(),
+                          _getNcclDataType(data[i].type().scalarType()),
+                          ncclOp[operation],
+                          dstRank * data.size(),
+                          (*comms)[i],
+                          stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
+}
+
+void DataChannelNccl::reduce(at::Tensor& data,
+                             THDReduceOp operation,
+                             rank_type dstRank,
+                             THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec = {data};
+  reduce(dataVec, operation, dstRank, groupId);
+}
+
+void DataChannelNccl::broadcast(std::vector<at::Tensor>& data,
+                                rank_type srcRank,
+                                THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  // Check the tensor vector for consistency
+  if (!_tensorCheckHelper(data, data)) {
+    return;
+  }
+  _checkGroupIdValid(groupId);
+
+  auto ncclResourcePair = _getNcclResourcePair(data, groupId);
+  auto comms = ncclResourcePair.first;
+  auto events = ncclResourcePair.second;
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  std::unique_lock<std::mutex> cudaFreeMutexLock(
+      *(THCCachingAllocator_getCudaFreeMutex()));
+
+  NCCL_CHECK(ncclGroupStart());
+  for (size_t i = 0; i < data.size(); ++i) {
+
+    gpuGuard.setDevice(data[i].get_device());
+    auto stream = THCState_getCurrentStream(THDGetCudaState());
+
+    NCCL_CHECK(ncclBcast(data[i].data_ptr(),
+                         data[i].numel(),
+                         _getNcclDataType(data[i].type().scalarType()),
+                         srcRank * data.size(),
+                         (*comms)[i],
+                         stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
+}
+
+
+void DataChannelNccl::broadcast(at::Tensor& data,
+                                rank_type srcRank,
+                                THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec = {data};
+  broadcast(dataVec, srcRank, groupId);
+}
+
+
+void DataChannelNccl::barrier(THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  _checkGroupIdValid(groupId);
+
+  if (_groupNcclResources.find(groupId) == _groupNcclResources.end() ||
+      _groupDevices.find(groupId) == _groupDevices.end()) {
+    return;
+  }
+
+  auto devices = getDevicesList(_groupDevices[groupId]);
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  int idx = 0;
+  // Synchronize on the CUDA events
+  for (auto& event : *(_groupNcclResources[groupId].ncclCudaEvents())) {
+    gpuGuard.setDevice(devices[idx++]);
+    THCudaCheck(cudaEventSynchronize(event));
+  }
+}
```

Comment:
As I said before that's not a distributed barrier. E.g. if you never do any NCCL calls, it will be a no-op (and barriers are never no-ops!). Consider this code:
```python
print("a")
dist.barrier()
print("b")
```
If you used MPI, you would have a guarantee that all processes will print a before any will print b. This function doesn't have the same guarantees.

I don't think you need to keep these events. Just make these functions always queue work on whatever is current stream, and keep the synchronization up to the caller (just like we do in Gloo channel)

Reply:
Yes, this is not a general-purpose barrier, it's just a barrier for NCCL. I still think this function would be a good option for the user and would rather not let user to simple to cudaStreamSynchronize on the default stream if they ever need to wait.  I am gonna make this function a new function, something called ncclBarrier()

If you don't

Reply:
It's not, but semantically that's what it is supposed to mean. Not raising an error means breaking the contract of the data channel interface which is a bad thing. I think stream synchronization shouldn't be part of DataChannel. It's up to the user to ensure that streams they use wait for other streams (maybe they will want to sync with a stream different than the default one?). We already have code for this in DistributedDataParallel, so it doesn't add much

Reply:
Raise an error instead in the new code change